### ISSUE TITLE
Issue in do_action implementation for twig

### ISSUE THREAD

**Issue by nkmaurya-dev** [timber/timber#1528]
Hi Team, 

I have found that the **Timber is passing context** by default to every **do_action hook**. 

So if any callback function has any parameter or first parameter in case of multiple , that would be replaced as Timber context. It must be not happen, because any plugin or theme would have written them for their own use.

The timber is adding the **action** function to the **Twig** and passing the context there. I think it should not or in some way, if we can override that method to not the pass the context to any hook by default.

Please see the method add_timber_functions line: 37 and below is the do_action hook code snippet:
See: https://github.com/timber/timber/blob/master/lib/Twig.php

```php
$twig->addFunction(new Twig_Function('action', function( $context ) {
	$args = func_get_args();
	array_shift($args);
	$args[] = $context;
	call_user_func_array('do_action', $args);
}, array('needs_context' => true)));
```

For testing , if I did comment the line: 42, it means no context is passing to the callback functions and everything works well.

I'm explaining here the detailed explanation of code flow. Its just an example of that issue.

In tease-product.twig, I have used the hook as:

`{% do action( 'woocommerce_after_shop_loop_item') %}`

and the plugin [YITH WooCommerce Quick View](https://wordpress.org/plugins/yith-woocommerce-quick-view/) used this hook in their class as:

```
add_action( 'woocommerce_after_shop_loop_item', array( $this, 'yith_add_quick_view_button' ), 15 );
```

where `$this` : current class object
`yith_add_quick_view_button` : the callback function

So function declaration is:

```php
public function yith_add_quick_view_button( $product_id = 0, $label = '', $return = false ) {
	global $product;
                
	// get product id
	! $product_id && $product_id = yit_get_prop( $product, 'id', true );
	// get label
	! $label && $label = $this->get_button_label();

	$button = '<a href="#" class="button yith-wcqv-button" data-product_id="' . $product_id . '">' . $label . '</a>';
    $button = apply_filters( 'yith_add_quick_view_button_html', $button, $label, $product );

	if( $return ) {
		return $button;
	}

	echo $button;
}
```

Now, the first parameter i.e $product_id = 0 is automatically changed to Timber context, instead it should be 0 , if no value is passed for first parameter. Timber context is an array so it is causing the error in button HTML reasonably. 

We can't force to pass the context as argument to do_action callback function. 

Please suggest me here.

WordPress 4.8.1, 
PHP 7.1, 
Timber 1.4.1 (installed via WordPress.org)

Thanks,
Neeraj Maurya

**Comment by gchtr** [comment]
I see where the problem lies, I just ran into something similar myself. 

You’d have to call the action with the default arguments the `yith_add_quick_view_button()` function accepts:

```twig
{% do action('woocommerce_after_shop_loop_item', 0, '', false) %}
```

However, when the action is registered like the following, then only **1 argument** will be passed to the function.

```php
add_action(
    'woocommerce_after_shop_loop_item',
    array( $this, 'yith_add_quick_view_button' ),
    15
);
```

For the function `yith_add_quick_view_button()` to be able to accept all three arguments (`$product_id`, `$label` and `$return`), the number of arguments should be passed as a 4th argument:

```php
add_action(
    'woocommerce_after_shop_loop_item',
    array( $this, 'yith_add_quick_view_button' ),
    15, 3
);
```

But that’s probably a second issue not related to Timber directly.

So, I agree with you that maybe it’s not such a good idea **to pass the context to the callback action**, because of the following reasons:

- When no arguments are passed, the default arguments for a callback function won’t kick in. They can’t be used when they are overwritten by the context variable.
- When calling `do_action`, Timber doesn’t know which functions will be called and how many arguments they will receive. So it can’t always send context as the last argument.
- Oftentimes actions are registered by plugins, where developers have no easy access to adapt the number of arguments that are passed to a callback function registered with `add_action()`. They’d have to first remove the action and then reregister it, which is kind of a big hassle for functionality that is used so rarely.
- If we’d still need the default context in a callback function, we could use `$context = Timber::get_context()`. Because the context is cached, this shouldn’t have big performance implications.
- The only situation where I ever needed to use `{% do action() %}` was when working with plugins like WooCommerce. I think we should stick to the default way that WordPress works with actions. Otherwise we might run into other unexpected issues.

I’ll open a pull request for this.

**Comment by gchtr** [comment]
Going on from here, I see multiple options:

- Leave it as it is.
- Deprecate that functionality in an upcoming release.
- Deprecate that functionality in Timber 2.x, and until then, deactivate the functionality by default but add a filter to enable it when it should be needed. I’m thinking of this:

**functions.php**

```php
add_filter( 'timber/compatibility/add_context_to_actions', '__return_true' );
```
@jarednova What’s your opinion on this? Do you remember a specific use case for adding this in <https://github.com/timber/timber/commit/a0b25bea6839fb7ba6687647a974c63218c321c5>?

**Comment by jarednova** [comment]
@gchtr I believe this was closed by #1773; can you confirm?

**Comment by gchtr** [comment]
Oh yes, that pull request was specifically for this ticket. I’m closing this.
